google/cloud/bigquery/_tqdm_helpers.py

```diff
@@ -0,0 +1,105 @@
+    "A progress bar was requested, but there was an error loading the tqdm "
+    "library. Please install tqdm to use the progress bar functionality."
+)
+
+_PROGRESS_BAR_UPDATE_INTERVAL = 0.5
+
+
+def get_progress_bar(progress_bar_type, description, total, unit):
+    """Construct a tqdm progress bar object, if tqdm is     ."""
+    if tqdm is None:
+        if progress_bar_type is not None:
+            warnings.warn(_NO_TQDM_ERROR, UserWarning, stacklevel=3)
+        return None
+
+    try:
+        if progress_bar_type == "tqdm":
+            return tqdm.tqdm(desc=description, total=total, unit=unit)
+        elif progress_bar_type == "tqdm_notebook":
+            return tqdm.tqdm_notebook(desc=description, total=total, unit=unit)
+        elif progress_bar_type == "tqdm_gui":
+            return tqdm.tqdm_gui(desc=description, total=total, unit=unit)
+    except (KeyError, TypeError):
+        # Protect ourselves from any tqdm errors. In case of
+        # unexpected tqdm behavior, just fall back to showing
+        # no progress bar.
+        warnings.warn(_NO_TQDM_ERROR, UserWarning, stacklevel=3)
+    return None
+
+
+def wait_for_query(query_job, progress_bar_type=None):
+    """Return query result and display a progress bar while the query running, if tqdm is installed."""
+    if progress_bar_type is None:
+        query_result = query_job.result()
+    else:
+        start_time = time.time()
+        progress_bar = get_progress_bar(
+            progress_bar_type, "Query is running", 1, "query"
+        )
+        i = 0
+        while True:
+            if query_job.query_plan:
+                total = len(query_job.query_plan)
+                query_job.reload()  # Refreshes the state via a GET request.
+                current_stage = query_job.query_plan[i]
+                progress_bar.total = len(query_job.query_plan)
+                progress_bar.set_description(
+                    "Query executing stage {} and status {} : {:0.2f}s".format(
+                        current_stage.name,
+                        current_stage.status,
+                        time.time() - start_time,
+                    ),
+                )
+                try:
+                    query_result = query_job.result(
+                        timeout=_PROGRESS_BAR_UPDATE_INTERVAL
+                    )
+                    progress_bar.update(total)
+                    progress_bar.set_description(
+                        "Query complete after {:0.2f}s".format(
+                            time.time() - start_time
+                        ),
+                    )
+                    break
+                except concurrent.futures.TimeoutError:
+                    if current_stage.status == "COMPLETE":
+                        if i < total - 1:
+                            progress_bar.update(i + 1)
+                            i += 1
+                    continue
+            else:
+                query_result = query_job.result()
```

Comment:
We need a timeout here. It's not clear to me why the above `try` block is even in the `if` statement.

Reply:
Looks like the only difference is the presence of `total`, which could be set to a default value in the case where `query_plan` is not present.